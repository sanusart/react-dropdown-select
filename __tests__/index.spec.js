/**
 * @jest-environment jsdom
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import Select from '../src/index';

jest.useFakeTimers();

// Mock window methods
window.addEventListener = jest.fn();
window.removeEventListener = jest.fn();
Element.prototype.getBoundingClientRect = () => ({
  width: 100,
  height: 100,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0
});

const mockProps = (props = {}) => ({
  values: [],
  options: [
    { value: 1, label: 'Option 1' },
    { value: 2, label: 'Option 2' },
    { value: 3, label: 'Option 3', disabled: true }
  ],
  onChange: jest.fn(),
  ...props
});

const selectWithProps = (component) => {
  return TestRenderer.create(component, {
    createNodeMock: (element) => {
      if (element.type === 'div') {
        return {
          blur: jest.fn(),
          focus: jest.fn(),
          getBoundingClientRect: () => ({
            width: 100,
            height: 100,
            top: 0,
            left: 0,
            bottom: 0,
            right: 0
          })
        };
      }
      return null;
    }
  });
};

describe('<Select/> component', () => {
  beforeEach(() => {
    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const tree = selectWithProps(<Select {...mockProps()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles lifecycle methods correctly', () => {
    const { unmount } = render(<Select {...mockProps()} />);

    expect(window.addEventListener).toHaveBeenCalledWith('resize', expect.any(Function));
    expect(window.addEventListener).toHaveBeenCalledWith('scroll', expect.any(Function));

    unmount();
    expect(window.removeEventListener).toHaveBeenCalledWith('resize', expect.any(Function));
    expect(window.removeEventListener).toHaveBeenCalledWith('scroll', expect.any(Function));
  });

  it('handles dropdown open/close', async () => {
    const { container } = render(<Select options={mockProps().options} />);
    const selectNode = container.querySelector('.react-dropdown-select');

    act(() => {
      fireEvent.mouseDown(selectNode);
      jest.runAllTimers();
    });

    await waitFor(() => {
      expect(selectNode.getAttribute('aria-expanded')).toBe('true');
      expect(container.querySelector('.react-dropdown-select-dropdown')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.mouseDown(document.body);
      jest.runAllTimers();
    });

    await waitFor(() => {
      expect(selectNode.getAttribute('aria-expanded')).toBe('false');
      expect(container.querySelector('.react-dropdown-select-dropdown')).not.toBeInTheDocument();
    });
  });

  it('handles item selection in multi mode', () => {
    const onChange = jest.fn();
    const component = selectWithProps(<Select {...mockProps({ onChange, multi: true })} />);

    TestRenderer.act(() => {
      component.toTree().instance.addItem({ value: 1, label: 'Option 1' });
    });

    expect(component.toTree().instance.state.values).toHaveLength(1);
    expect(onChange).toHaveBeenCalled();

    TestRenderer.act(() => {
      component.toTree().instance.removeItem(null, { value: 1, label: 'Option 1' });
    });

    expect(component.toTree().instance.state.values).toHaveLength(0);
  });

  it('handles search functionality', () => {
    const component = selectWithProps(<Select {...mockProps()} />);

    TestRenderer.act(() => {
      component.toTree().instance.setSearch({ target: { value: 'Option 1' } });
    });

    expect(component.toTree().instance.state.search).toBe('Option 1');
    expect(component.toTree().instance.state.searchResults).toHaveLength(1);
  });

  it('handles select all functionality', () => {
    const component = selectWithProps(<Select {...mockProps({ multi: true })} />);

    TestRenderer.act(() => {
      component.toTree().instance.selectAll();
    });

    expect(component.toTree().instance.state.values).toHaveLength(2); // Excluding disabled option
  });

  it('handles clear all functionality', () => {
    const component = selectWithProps(
      <Select
        {...mockProps({
          multi: true,
          values: [{ value: 1, label: 'Option 1' }, { value: 2, label: 'Option 2' }]
        })}
      />
    );

    TestRenderer.act(() => {
      component.toTree().instance.clearAll();
    });

    expect(component.toTree().instance.state.values).toHaveLength(0);
  });

  it('handles keyboard navigation', () => {
    const { container } = render(<Select {...mockProps()} />);

    // Open dropdown
    act(() => {
      fireEvent.mouseDown(container.querySelector('.react-dropdown-select'));
    });

    // Press down arrow
    act(() => {
      fireEvent.keyDown(container.querySelector('.react-dropdown-select'), { key: 'ArrowDown' });
    });

    const options = container.querySelectorAll('.react-dropdown-select-item');
    expect(options[0]).toHaveClass('react-dropdown-select-item-active');
  });

  it('handles create new option', () => {
    const component = selectWithProps(
      <Select
        {...mockProps({
          create: true,
          createNewLabel: 'Create: '
        })}
      />
    );

    TestRenderer.act(() => {
      component.toTree().instance.setSearch({ target: { value: 'New Option' } });
      component.toTree().instance.createNew('New Option');
    });

    expect(component.toTree().instance.state.values[0].label).toBe('New Option');
  });

  it('handles closeOnSelect prop', () => {
    const component = selectWithProps(<Select {...mockProps({ closeOnSelect: true })} />);

    TestRenderer.act(() => {
      component.toTree().instance.dropDown('open');
      component.toTree().instance.addItem({ value: 1, label: 'Option 1' });
    });

    expect(component.toTree().instance.state.dropdown).toBe(false);
  });

  it.skip('handles closeOnScroll prop', async () => {
    jest.useFakeTimers();

    const onDropdownCloseRequest = jest.fn(({ close }) => {
      close();
    });

    const { container } = render(
      <Select
        options={mockProps().options}
        closeOnScroll
        onDropdownCloseRequest={onDropdownCloseRequest}
        debounceDelay={100}
      />
    );

    const selectNode = container.querySelector('.react-dropdown-select');

    // First, ensure the dropdown is open
    act(() => {
      fireEvent.mouseDown(selectNode);
      jest.runAllTimers();
    });

    await waitFor(() => {
      expect(selectNode.getAttribute('aria-expanded')).toBe('true');
      expect(container.querySelector('.react-dropdown-select-dropdown')).toBeInTheDocument();
    });

    // Now trigger the scroll event
    act(() => {
      const scrollEvent = new Event('scroll');
      window.dispatchEvent(scrollEvent);
      // Wait for the debounce delay
      jest.advanceTimersByTime(100);
      jest.runAllTimers();
    });

    // Wait for onDropdownCloseRequest to be called
    await waitFor(() => {
      expect(onDropdownCloseRequest).toHaveBeenCalledWith({
        props: expect.any(Object),
        methods: expect.any(Object),
        state: expect.objectContaining({
          dropdown: true
        }),
        close: expect.any(Function)
      });
    });

    // Get the close function and call it
    const closeFunction = onDropdownCloseRequest.mock.calls[0][0].close;
    act(() => {
      closeFunction();
      jest.runAllTimers();
    });

    // Verify the dropdown is closed
    await waitFor(() => {
      expect(selectNode.getAttribute('aria-expanded')).toBe('false');
      expect(container.querySelector('.react-dropdown-select-dropdown')).not.toBeInTheDocument();
    });

    jest.useRealTimers();
  });

  it('handles custom compare function', () => {
    const compareValuesFunc = (a, b) => JSON.stringify(a) === JSON.stringify(b);
    const component = selectWithProps(<Select {...mockProps({ compareValuesFunc })} />);

    const newValues = [{ value: 1, label: 'Option 1' }];
    TestRenderer.act(() => {
      component.root.instance.componentDidUpdate({ ...mockProps(), values: [] }, { values: [] });
    });

    expect(component.toTree().instance.state.values).toEqual([]);
  });

  it('handles onDropdownCloseRequest prop', async () => {
    const onDropdownCloseRequest = jest.fn();
    const { container } = render(
      <Select options={mockProps().options} onDropdownCloseRequest={onDropdownCloseRequest} />
    );
    const selectNode = container.querySelector('.react-dropdown-select');

    act(() => {
      fireEvent.mouseDown(selectNode);
      jest.runAllTimers();
    });

    await waitFor(() => {
      expect(selectNode.getAttribute('aria-expanded')).toBe('true');
      expect(container.querySelector('.react-dropdown-select-dropdown')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.mouseDown(document.body);
      jest.runAllTimers();
    });

    await waitFor(() => {
      expect(onDropdownCloseRequest).toHaveBeenCalled();
    });
  });
});
