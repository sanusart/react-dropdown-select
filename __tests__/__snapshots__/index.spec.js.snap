// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select/> component renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 1px solid #ccc;
  width: 100%;
  border-radius: 2px;
  padding: 2px 5px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  direction: ltr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  min-height: 36px;
  pointer-events: all;
}

.emotion-0:hover,
.emotion-0:focus-within {
  border-color: #0074D9;
}

.emotion-0:focus,
.emotion-0:focus-within {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(0, 116, 217, 0.2);
}

.emotion-0 * {
  box-sizing: border-box;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-4 {
  line-height: inherit;
  border: none;
  margin-left: 5px;
  background: transparent;
  padding: 0;
  width: calc(9ch + 5px);
  font-size: smaller;
}

.emotion-4:focus {
  outline: none;
}

.emotion-6 {
  text-align: center;
  pointer-events: none;
  margin: 0 0 0 5px;
  -webkit-transform: rotate(180deg);
  -moz-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
  cursor: pointer;
}

.emotion-6 svg {
  width: 16px;
  height: 16px;
}

.emotion-6:hover path {
  stroke: #0074D9;
}

.emotion-6:focus {
  outline: none;
}

.emotion-6:focus path {
  stroke: #0074D9;
}

<div>
  <div
    aria-expanded={false}
    aria-label="Dropdown select"
    className="react-dropdown-select undefined emotion-0 emotion-1"
    color="#0074D9"
    direction="ltr"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex="0"
  >
    <div
      className="react-dropdown-select-content react-dropdown-select-type-single emotion-2 emotion-3"
      onClick={[Function]}
    >
      <input
        className="react-dropdown-select-input emotion-4 emotion-5"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        placeholder="Select..."
        readOnly={false}
        size={9}
        tabIndex="-1"
        value=""
      />
    </div>
    <div
      className="react-dropdown-select-dropdown-handle emotion-6 emotion-7"
      color="#0074D9"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      rotate={1}
      tabIndex="-1"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 40 40"
      >
        <path
          d="M31 26.4q0 .3-.2.5l-1.1 1.2q-.3.2-.6.2t-.5-.2l-8.7-8.8-8.8 8.8q-.2.2-.5.2t-.5-.2l-1.2-1.2q-.2-.2-.2-.5t.2-.5l10.4-10.4q.3-.2.6-.2t.5.2l10.4 10.4q.2.2.2.5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
